### PR TITLE
Add detection confidence curves

### PR DESCRIPTION
**Pull request by rolandocortez**
# Add precision-recall and confidence curves for detection

## Purpose

Add confidence-based evaluation curves for bounding-box detection in `luxonis-train`.

The new metric provides:

* precision-recall
* precision-confidence
* recall-confidence

The current detection metrics operate on post-NMS detections produced at a fixed confidence threshold. That is sufficient for scalar metrics such as mAP, but it does not provide the decoded candidates needed to evaluate how precision and recall change across confidence thresholds.

This PR adds an opt-in `PrecisionRecallCurve` metric and exposes decoded pre-NMS detections from the supported bounding-box heads during evaluation.

### Detection-head output

Updated:

* `EfficientBBoxHead`
* `PrecisionBBoxHead`

Both heads now expose:

* the existing post-NMS `boundingbox` output
* a new decoded `detections_pre_nms` tensor during evaluation

Training and export output contracts remain unchanged.

For `EfficientBBoxHead`, decoding and NMS are separated into:

* `_prepare_bbox_inference_output()`
* `_run_nms()`

The existing `_postprocess_detections()` behavior is preserved as a wrapper around those two steps.

### PrecisionRecallCurve metric

Added an opt-in `PrecisionRecallCurve` metric for bounding-box detection.

The metric:

* supports an explicit confidence grid or a generated linear grid
* runs NMS once at the minimum configured confidence threshold
* reuses the retained detections across the confidence grid
* performs class-aware, score-ordered, one-to-one matching
* uses a configurable matching IoU threshold
* defaults NMS IoU and maximum detections to the attached detection head values
* accumulates true positives, false positives, and target counts with distributed sum reduction

The metric computes:

* precision for each confidence threshold
* recall for each confidence threshold
* F1 for each confidence threshold
* maximum F1
* confidence threshold at maximum F1

`max_f1` is exposed as the primary scalar under the metric name.

`confidence_at_max_f1` is exposed as a submetric.

### Metric artifacts

Extended the metric interface with generic hooks for:

* filtering values intended for scalar logging
* generating image artifacts from computed metric results
* declaring stable artifact names

`PrecisionRecallCurve` logs one RGB image containing:

* precision-recall
* precision-confidence
* recall-confidence

The artifact is generated only on global rank zero.

### Scope

This PR intentionally supports:

* bounding-box detection only
* global micro-averaged curves
* one matching IoU threshold
* one NMS pass at the minimum confidence threshold

The following are not included:

* per-class curves
* AUPRC
* instance-keypoint or segmentation curves

mAP remains the recommended primary detection metric.

## Dependencies & Potential Impact

No new external dependency is introduced.

Potential impact:

* evaluation packets from `EfficientBBoxHead` and `PrecisionBBoxHead` now include the additional `detections_pre_nms` key
* the existing post-NMS `boundingbox` output remains unchanged
* training output remains unchanged
* export output remains unchanged
* the curve metric is opt-in and does not affect configurations that do not attach it
* generating the curve artifact requires rendering one Matplotlib figure at evaluation epoch end on global rank zero

The metric uses the decoded candidates produced by the detection head, so the supported heads and the metric share the same pre-NMS tensor contract.

## Testing & Validation

Validation performed:

* Ruff checks passed
* formatting checks passed
* Python compilation passed
* `git diff --check` passed
* private metadata and temporary-code checks passed

Focused metric validation:

* 23 `PrecisionRecallCurve` tests passed
* invalid configuration and empty-input cases covered
* accumulation and reset behavior covered
* scalar and artifact separation covered
* behavior with `log_sub_metrics=True` and `False` covered
* MLflow metric and artifact keys covered

Detection-head validation:

* real forward-path tests for `EfficientBBoxHead`
* real forward-path tests for `PrecisionBBoxHead`
* exact replay of the canonical post-NMS output from `detections_pre_nms`
* training contract checked
* evaluation contract checked
* export contract checked

Regression validation:

* 95 metric unit tests passed
* 4 Lightning utility tests passed

Distributed validation:

* two-process CPU/Gloo validation passed
* distributed states matched a single-process reference
* computed vectors and primary scalar were identical across ranks
* curve artifact generation occurred only on rank zero

The final commit was also rebased onto the current `main` and revalidated before push.

## Deployment Plan

No deployment is required.

Rollout:

* review the implementation and metric contract
* verify hosted CI
* use `PrecisionRecallCurve` only in detection configurations that require confidence-based curves
* keep mAP as the default primary detection metric

Rollback:

* revert this PR if the additional evaluation packet output or metric artifact hooks cause regressions

Monitoring:

* hosted CI results
* reviewer feedback on the pre-NMS output contract
* reviewer feedback on the global micro matching and single-NMS design
* runtime and memory behavior in larger detection validation runs

## AI Usage

Assisted-by: ChatGPT

## Summary by CodeRabbit

- **New Features**
  - Added a bounding-box Precision/Recall curve metric with precision, recall, F1, and best-confidence reporting.
  - Added visual curve artifacts for validation and testing results.

- **Improvements**
  - Standardized metric logging and artifact handling.
  - Inference now exposes pre-NMS detections for more accurate evaluation.

- **Tests**
  - Added comprehensive metric, logging, artifact, and detection-head integration coverage.

- **Chores**
  - Added Matplotlib for curve rendering.

### CODE REVIEW

File: luxonis_train/lightning/luxonis_lightning.py [hunk 1]
```diff
@@ -878,6 +878,36 @@ def _evaluation_epoch_end(self, mode: Literal["test", "val"]) -> None:
...
+                        self.tracker.log_image(
+                            name=f"{mode}/metrics/{self.current_epoch}/"
+                            f"{formatted_node_name}/{metric_name}/"
+                            f"{artifact_name}",
+                            img=artifact.detach()
+                            .cpu()
+                            .numpy()
+                            .transpose(1, 2, 0),
+                            step=self.current_epoch,
+                        )
```
[the remark this answers is not in the document]
Comment (rolandocortez) [reply luxonis_train/lightning/luxonis_lightning.py]: Addressed in the latest update. Synchronized scalar logging now completes before rank-zero artifact generation. Artifact generation errors and invalid shapes are handled with warning-and-skip behavior instead of raising on a single rank. I also added regression tests for both failure paths

File: luxonis_train/attached_modules/metrics/precision_recall_curve.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        """Return the stable name of the combined curve figure."""
+        return ("curves",)
+
+    def get_artifacts(
+        self,
+        values: dict[str, Tensor],
+    ) -> dict[str, Tensor]:
+        """Render PR, precision-confidence, and recall-confidence curves."""
+        from matplotlib.backends.backend_agg import FigureCanvasAgg
+        from matplotlib.figure import Figure
```
[the remark this answers is not in the document]
Comment (rolandocortez) [reply luxonis_train/attached_modules/metrics/precision_recall_curve.py:352]: Addressed in the latest update. Matplotlib is now declared directly as a runtime dependency because the metric imports it for curve artifact rendering. It was previously available transitively through Seaborn

File: luxonis_train/attached_modules/metrics/precision_recall_curve.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+        @type matching_iou_threshold: float
+        @param matching_iou_threshold: IoU required for a prediction to match
+            a ground-truth box.
+        @type nms_iou_threshold: float | None
+        @param nms_iou_threshold: IoU used by NMS. Defaults to the attached
+            detection head value.
+        @type max_detections: int | None
+        @param max_detections: Maximum detections retained by NMS. Defaults to
+            the attached detection head value.
+        """
```
[the remark this answers is not in the document]
Comment (rolandocortez) [reply luxonis_train/attached_modules/metrics/precision_recall_curve.py:54]: Addressed in the latest update. I ran the full pre-commit suite and applied the docformatter changes. All pre-commit hooks now pass locally

File: luxonis_train/lightning/luxonis_lightning.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rolandocortez) [reply luxonis_train/lightning/luxonis_lightning.py]: Addressed in [16c1be1](https://github.com/luxonis/luxonis-train/pull/429/commits/16c1be14781aa571bc616e5b9bd125e004f43e81). The artifact tensor conversion and tracker.log_image() call are now wrapped in the rank-zero fail-soft path, so conversion or tracker backend failures are logged and skipped instead of raising on a single rank. I also added a regression test covering a log_image() failure while confirming that scalar logging and metric reset still complete

File: luxonis_train/attached_modules/metrics/base_metric.py [hunk 4]
```diff
@@ -165,6 +167,29 @@ def compute(
...
+        return values
+
+    def get_artifacts(
+        self,
+        values: MetricResult,
+    ) -> dict[str, Tensor]:
+        """Build image artifacts from an already computed metric
+        result.
+        """
+        del values
```
Comment (kozlov721) [inline luxonis_train/attached_modules/metrics/base_metric.py]: Why do we delete the values here? It looks like a possibly dangerous side effect to me.

Comment (rolandocortez) [reply luxonis_train/attached_modules/metrics/base_metric.py]: Fair question. In this case, `del values` only removes the local parameter binding and does not mutate the metric result owned by the caller. It was used to make the intentionally unused argument explicit in the default implementation, but since the method simply returns an empty artifact mapping, leaving the parameter untouched is clearer. I’ll remove the line

Comment (rolandocortez) [reply luxonis_train/attached_modules/metrics/base_metric.py]: Adressed in [825c06d](https://github.com/luxonis/luxonis-train/pull/429/commits/825c06d4b9330c8d7e5935d20afaf0c3508bddeb)

**Review by kozlov721 — APPROVED** [review]

**Review by klemen1999 — APPROVED** [review]
